tools/contractdeployer/contract_deployer.go

```diff
@@ -141,67 +117,14 @@ func setupWallet(cfg *Config) (wallet.Wallet, error) {
 	return wallet.NewInMemoryWalletFromPK(cfg.ChainID, privateKey), nil
 }
 
-func getClient(config *Config, wal wallet.Wallet) (ethadapter.EthClient, error) {
-	var client ethadapter.EthClient
-	var err error
-
-	startConnectingTime := time.Now()
-	// since the nodes we are connecting to may have only just started, we retry connection until it is successful
-	for client == nil && time.Since(startConnectingTime) < timeoutWait {
-		client, err = setupClient(config, wal)
-		if err == nil {
-			break // success
-		}
-		// if there was an error we'll retry, if we timeout the last seen error will display
-		time.Sleep(retryInterval)
-	}
-	if client == nil {
-		return nil, fmt.Errorf("failed to initialise client connection after retrying for %s, %w", timeoutWait, err)
-	}
-
-	return client, nil
-}
-
-func setupClient(cfg *Config, wal wallet.Wallet) (ethadapter.EthClient, error) {
-	if cfg.IsL1Deployment {
-		// return a connection to the l1
-		return ethadapter.NewEthClient(cfg.NodeHost, cfg.NodePort, 30*time.Second, common.HexToAddress("0x0"))
-	}
-	return ethadapter.NewObscuroRPCClient(cfg.NodeHost, cfg.NodePort, wal)
-}
-
-func (cd *contractDeployer) prefundAccount() error {
-	balance, err := cd.client.BalanceAt(cd.wallet.Address(), nil)
-	if err != nil {
-		return fmt.Errorf("failed to fetch contract deployer balance. Cause: %w", err)
-	}
-	// We do not send more funds if the contract deployer already has enough.
-	if balance.Cmp(big.NewInt(Prealloc)) == 1 {
-		return nil
-	}
-
-	nonce, err := cd.faucetClient.Nonce(cd.faucetWallet.Address())
-	if err != nil {
-		return fmt.Errorf("failed to fetch faucet nonce. Cause: %w", err)
+func prepareClient(config *Config, wal wallet.Wallet) (contractDeployerClient, error) {
+	if config.IsL1Deployment {
+		return prepareEthDeployer(config)
 	}
-
-	destAddr := cd.wallet.Address()
-	tx := &types.LegacyTx{
-		Nonce:    nonce,
-		Value:    big.NewInt(Prealloc),
-		Gas:      uint64(1_000_000),
-		GasPrice: common.Big1,
-		To:       &destAddr,
-	}
-	_, err = signAndSendTxWithReceipt(cd.faucetWallet, cd.faucetClient, tx)
-	if err != nil {
-		return fmt.Errorf("failed to complete faucet transaction. Cause: %w", err)
-	}
-
-	return nil
+	return prepareObscuroDeployer(config, wal)
 }
 
-func signAndSendTxWithReceipt(wallet wallet.Wallet, client ethadapter.EthClient, tx *types.LegacyTx) (*common.Address, error) {
+func signAndSendTxWithReceipt(wallet wallet.Wallet, client contractDeployerClient, tx *types.LegacyTx) (*common.Address, error) {
```

Comment:
client -> deployer

(client is used for the underlying things as well)

Reply:
Below in this function there is logic to wait for the receipt.
Couldn't that use the new function?

Reply:
It can't be easily re-used here unfortunately, because this code has to work for both the L1 deployer and the L2 deployer. But I'll make sure that other function gets re-used in follow-up PRs, there's a bunch of scripts and tests where we end up re-writing the same retry/waiting code I think